src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -268,6 +274,33 @@ private void CheckHeaderSignatureWithCoinstakeKernel(ProvenBlockHeader header, U
             ConsensusErrors.BadBlockSignature.Throw();
         }
 
+        /// <summary>
+        /// Checks if coinstake is spent on another chain.
+        /// </summary>
+        /// <param name="header">The proven block header.</param>
+        /// <exception cref="UtxoNotFoundInRewindDataException">
+        /// Throws specific exception so that it can be handled later.
+        /// </exception>
+        private void CheckIfCoinstakeIsSpentOnAnotherChain(ProvenBlockHeader header)
+        {
+            Transaction coinstake = header.Coinstake;
+            TxIn intput = coinstake.Inputs[0];
+
+            int? rewindDataIndex = this.PosParent.RewindDataIndexStore.GetAsync(intput.PrevOut.Hash, (int)intput.PrevOut.N).GetAwaiter().GetResult();
```

Comment:
I'm not sure because I've to re-read the requirement, but it seems to me that we have to check in all RewindData we stored between new chain tip and fork point, so we have to fetch from RewindDataIndexStore an index to the RewindData next of the fork point  between the new chain and the current chain
Using the coinstake transaction of fork point as TX+N we should get the index of the RewindData and we need to check that from that RewindData to the last(?) our utxo isn't spent.

I may be wrong tho, can anyone confirm?

Reply:
@dangershony Can you have a look, please. This is important.